<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Lederhosen by audy</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>Lederhosen</h1>
        <p>Cluster raw Illumina 16S rRNA amplicon data to generate OTUs.</p>

        <p class="view"><a href="https://github.com/audy/lederhosen">View the Project on GitHub <small>audy/lederhosen</small></a></p>


        <ul>
          <li><a href="https://github.com/audy/lederhosen/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/audy/lederhosen/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/audy/lederhosen">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h1>Lederhosen</h1>

<p>Cluster raw Illumina 16S rRNA amplicon data to generate OTUs.</p>

<h3>About</h3>

<ul>
<li>Lederhosen is a project born out of the Triplett Lab at the University of Florida.</li>
<li>Lederhosen is designed to be a fast and simple method of clustering 16S rRNA amplicons sequenced
using paired and non-paired end short reads such as those produced by Illumina (GAIIx, HiSeq and MiSeq).</li>
<li>Lederhosen uses <a href="http://semver.org/">Semantic Versioning</a>.</li>
<li>Lederhosen is free and open source under the <a href="http://opensource.org/licenses/mit-license.php/">MIT open source license</a>.</li>
<li>Except for USEARCH which requires a license, Lederhosen is available for commercial use.</li>
</ul><h2>How do I get Lederhosen?</h2>

<ol>
<li>Obtain &amp; Install <a href="http://www.drive5.com/">USEARCH</a> (32bit is fine for non-commercial use)</li>
<li>Get a copy of <a href="http://github.com/audy/taxcollector">TaxCollector</a>
</li>
<li>
<p>Install Lederhosen by typing:</p>

<p><code>sudo gem install lederhosen</code></p>
</li>
<li><p>Check installation by typing <code>lederhosen</code>. You should see some help text.</p></li>
</ol><h2>Features</h2>

<ul>
<li>Sequence trimming (paired-end Illumina).</li>
<li>Parallel, referenced-based clustering to TaxCollector using USEARCH</li>
<li>Generation and filtering of OTU abundancy matrices.</li>
</ul><h2>How do I use Lederhosen?</h2>

<p>Lederhosen is just a convenient wrapper for UCLUST and BLAT with some scripts for quality filtering, de-noising of data as well as creation of nice tables. It is similar to QIIME but meant for paired-end Illumina data rather than single-end 454. The basic lederhosen pipeline consists of: trimming, joining, sorting, filtering, clustering, more filtering, and output generation (OTU tables, representative reads, reads by cluster, and taxonomic descriptions for clusters). See the example pipeline in <code>pipeline.sh</code>.</p>

<h2>Tasks</h2>

<p>Lederhosen is invoked by typing <code>lederhosen [TASK]</code></p>

<h3>Trim Reads</h3>

<p>Trim (Illumina) reads using quality scores. Output will be a directory of fasta files. Reads can optionally be gzipped.</p>

<pre><code>lederhosen trim --reads_dir=reads/*.txt --out_dir=trimmed/
</code></pre>

<p>The trimming process will reverse complement the "right" pair so that both reads are in the forward orientation.</p>

<h3>Create Database</h3>

<p>Create UDB database required by usearch from TaxCollector</p>

<pre><code>lederhosen make_udb --input=taxcollector.fa --output=taxcollector.udb
</code></pre>

<h3>Cluster Reads using USEARCH</h3>

<p>Cluster reads using USEARCH. Output is a uc file.</p>

<pre><code>lederhosen cluster --input=trimmed/*.fasta --identity=0.95 --output=clusters_95.uc --database=taxcollector.udb
</code></pre>

<h3>Generate OTU table(s)</h3>

<p>Create an OTU abundance table where rows are samples and columns are clusters. The entries are the number of reads for that cluster in a sample.</p>

<pre><code>lederhosen otu_table --clusters=clusters_95.uc --prefix=otu_table --level=domain phylum class order family genus species
</code></pre>

<p>This will create the files:</p>

<pre><code>otu_table.domain.csv, ..., otu_table.species.csv
</code></pre>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/audy">audy</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>